csrc/preseg_passes/fmin_fmax_promotion.cpp

```diff
@@ -0,0 +1,298 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/fmin_fmax_promotion.h>
+
+#include <unordered_map>
+#include <vector>
+
+#include <ir/utils.h>
+#include <logical_domain_map.h>
+
+namespace nvfuser::preseg_passes {
+
+// IterDomainStatus are attached to IterDomains and propagated with a
+// downward-flow algorithm. The goal is to detect whether lost NANs propagated
+// to the outputs of a fusion. If so, then it is not valid to do the fast
+// min/max promotion.
+//
+// "BAD" statuses indicate ID's that potentially lost their NANs.
+// "GOOD" statuses will repair BAD statuses.
+// "REDUCE" statuses are reduced dimensions, and should never be mapped to ID's
+//   with non-reduced status.
+// All other statuses are "full-sized" (besides NONE which says nothing).
+//
+// A reduction ID will convert a full-size status into a REDUCE status. Likewise
+// a broadcast ID will convert a reduced status to a full size status.
+//
+// Statuses can interact via binary ops in a kind of algebra. The two reduced
+// statuses have a simple interaction: GOOD_REDUCE beats BAD-REDUCE.
+//
+// The 4 full-sized statuses have a slightly more complicated interaction:
+// 1. Any status matched with itself, produces itself.
+// 2. GOOD_BROADCAST matched with anything produces GOOD_BROADCAST.
+// 3. All other cases produce BAD_BROADCAST_DEFAULT
+//
+// An example of what each status means is below (with *max* reduction):
+// [0.0  1.0  2.0  3.0  NAN  5.0] <- DEFAULT
+// [5.0]                          <- BAD_REDUCE
+// [NAN]                          <- GOOD_REDUCE
+// [5.0  5.0  5.0  5.0  5.0  5.0] <- BAD_BROADCAST
+// [5.0  5.0  5.0  5.0  NAN  5.0] <- BAD_BROADCAST_DEFAULT
+// [NAN  NAN  NAN  NAN  NAN  NAN] <- GOOD_BROADCAST
+enum class IterDomainStatus {
+  // NONE is the status when hitting untracked ID's.
+  NONE,
+
+  // Reduced statuses
+  BAD_REDUCE,
+  GOOD_REDUCE,
+
+  // Full-size statuses
+  DEFAULT,
+  BAD_BROADCAST,
+  BAD_BROADCAST_DEFAULT,
+  GOOD_BROADCAST,
+};
+
+using IterStatusMap = std::unordered_map<IterDomain*, IterDomainStatus>;
+
+IterDomainStatus BopStatus(IterDomainStatus lhs, IterDomainStatus rhs) {
```

Comment:
nit: our function names start with a lower letter.

```suggestion
IterDomainStatus bopStatus(IterDomainStatus lhs, IterDomainStatus rhs) {
```